File: examples/java/sdk-reference/error-handling/with-retry-helper.java
import java.time.Duration;
import java.util.Map;
import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.DurableHandler;
import software.amazon.lambda.durable.config.WithRetryConfig;
import software.amazon.lambda.durable.retry.JitterStrategy;
import software.amazon.lambda.durable.retry.RetryStrategies;

public class WithRetryHelperExample extends DurableHandler<Map<String, Object>, String> {

    @Override
    public String handleRequest(Map<String, Object> event, DurableContext context) {
        WithRetryConfig retryConfig = WithRetryConfig.builder()
                .retryStrategy(RetryStrategies.exponentialBackoff(
                        3,                       // maxAttempts
                        Duration.ofSeconds(2),   // initialDelay
                        Duration.ofMinutes(1),   // maxDelay
                        2.0,                     // backoffRate
                        JitterStrategy.FULL))    // jitter
                .build();

        // invoke does not accept a retry strategy, so withRetry applies backoff
        // between failed attempts.
        return context.withRetry(
                "charge-payment",
                (attempt, ctx) -> ctx.invoke(
                        "charge-" + attempt,
                        "process-payment",
                        Map.of("orderId", event.get("orderId")),
                        String.class),
                retryConfig);
    }
}

File: examples/typescript/sdk-reference/error-handling/with-retry-helper.ts
import {
  withDurableExecution,
  withRetry,
  createRetryStrategy,
} from "@aws/durable-execution-sdk-js";

const retryStrategy = createRetryStrategy({
  maxAttempts: 3,
  initialDelay: { seconds: 2 },
  backoffRate: 2,
});

export const handler = withDurableExecution(async (event, context) => {
  // invoke does not accept a retryStrategy, so wrap it with withRetry to
  // apply backoff between failed attempts.
  const receipt = await withRetry(
    context,
    "charge-payment",
    (ctx, attempt) =>
      ctx.invoke<{ orderId: string }, string>(
        `charge-${attempt}`,
        "process-payment",
        { orderId: (event as { orderId: string }).orderId },
      ),
    { retryStrategy },
  );
  return receipt;
});
